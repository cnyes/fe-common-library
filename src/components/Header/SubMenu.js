/* eslint import/no-extraneous-dependencies: ["error", {"peerDependencies": true}] */
import React, { Component, PropTypes } from 'react';
import idx from 'idx';
import classNames from 'classnames';
import { requestType } from '../../utils/propTypes';
import getStyleName from '../../utils/getStyleName';
import styles from './Header.scss';
import NewsPlaceholder from './HeaderNavNewsPlaceholder/Row';

const EMPTY_ARRAY = [];

const itemType = PropTypes.shape({
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
});

const style = {
  narrowedPopup: {
    width: '120px',
    height: 'auto',
    overflow: 'visible',
  },
  narrowedList: {
    height: 'auto',
  },
};

function renderList(list, baseUrl, Link) {
  const content = (list || EMPTY_ARRAY).map(item => {
    return baseUrl.length || !Link ? (
      <a
        key={item.newsId}
        href={`${baseUrl}/news/id/${item.newsId}`}
        title={item.title}
        target="_blank"
        rel="noopener noreferrer"
      >
        {item.title}
      </a>
    ) : (
      <Link key={item.newsId} to={{ pathname: `/news/id/${item.newsId}` }} title={item.title}>
        {item.title}
      </Link>
    );
  });

  return (
    <nav className={classNames(getStyleName(styles, 'news-list'), 'theme-news-list')}>
      <h5>新聞頭條</h5>
      {!content.length ? <NewsPlaceholder /> : content}
    </nav>
  );
}

function renderStaticList(list, className, wrapperStyle, listTitle, Link) {
  const content = list.map(({ url, title, icon }, id) => {
    if (url && url[0] === '/') {
      return Link ? (
        <Link
          key={id} // eslint-disable-line react/no-array-index-key
          to={{ pathname: url }}
          className={icon ? getStyleName(styles, `icon-${icon}`) : null}
          title={title}
        >
          {title}
        </Link>
      ) : (
        <a
          key={id} // eslint-disable-line react/no-array-index-key
          href={url}
          className={icon ? getStyleName(styles, `icon-${icon}`) : null}
          title={title}
        >
          {title}
        </a>
      );
    }

    return (
      <a
        key={id} // eslint-disable-line react/no-array-index-key
        href={url}
        className={icon ? getStyleName(styles, `icon-${icon}`) : null}
        title={title}
      >
        {title}
      </a>
    );
  });

  return (
    <nav className={className} style={wrapperStyle || {}}>
      {listTitle && <h5>{listTitle}</h5>}
      {content}
    </nav>
  );
}

export default class SubMenu extends Component {
  static propTypes = {
    // ownProps
    /* tab label */
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    className: PropTypes.string,
    catSlug: PropTypes.string,
    /* Menu */
    leftList: PropTypes.arrayOf(itemType),
    leftListTitle: PropTypes.string,
    rightList: PropTypes.arrayOf(itemType),
    rightListTitle: PropTypes.string,
    // eslint-disable-next-line react/no-unused-prop-types
    newsBaseUrl: PropTypes.string.isRequired,
    request: requestType,
    Link: PropTypes.func,
  };

  static defaultProps = {
    className: '',
    catSlug: undefined,
    leftList: undefined,
    leftListTitle: undefined,
    rightList: undefined,
    rightListTitle: undefined,
    newsBaseUrl: '',
    request: undefined,
    Link: undefined,
  };

  constructor(...args) {
    super(...args);

    this.state = {
      list: [],
    };
  }

  loadNewsByCat = () => {
    const { catSlug, request } = this.props;

    if (!catSlug || !request) {
      return;
    }
    const api = `/api/v3/news/category/${catSlug}?isCategoryHeadline=1&limit=30`;

    request.get(api).then(
      response => {
        const list = idx(response, _ => _.data.items.data) || [];

        this.setState({ list });
      },
      error => {
        if (__DEVELOPMENT__) {
          console.error(error);
        }
      }
    );
  };

  renderPopupMenu() {
    const { leftList, leftListTitle, catSlug, rightList, rightListTitle, newsBaseUrl, request, Link } = this.props;
    const { list } = this.state;
    const isOnlyLeft = (!catSlug && !rightList) || (catSlug && !request);

    if (!leftList && isOnlyLeft) {
      // no left and no right, => no popup menu
      return null;
    }

    return (
      <div
        className={classNames(getStyleName(styles, 'popup'), 'theme-popup')}
        style={(isOnlyLeft && style.narrowedPopup) || {}}
      >
        {leftList &&
          renderStaticList(
            leftList,
            classNames(getStyleName(styles, 'link-wrapper'), 'theme-link-wrapper'),
            isOnlyLeft && style.narrowedList,
            leftListTitle,
            Link
          )}
        {catSlug && request && renderList(list, newsBaseUrl, Link)}
        {!catSlug &&
          rightList &&
          renderStaticList(
            rightList,
            classNames(getStyleName(styles, 'news-list'), 'theme-news-list'),
            undefined,
            rightListTitle,
            Link
          )}
      </div>
    );
  }

  render() {
    const { url, title, className, leftList, catSlug, rightList } = this.props;
    const isOnlyLeft = !catSlug && !rightList;

    return (
      <div className={className} onMouseEnter={this.loadNewsByCat}>
        {url.length ? (
          <a href={url} className={getStyleName(styles, 'submenu-title')}>
            {title}
            {(leftList || !isOnlyLeft) && <span className={getStyleName(styles, 'with-arrow')} />}
          </a>
        ) : (
          <span className={getStyleName(styles, 'submenu-title')}>
            {title}
            {(leftList || !isOnlyLeft) && <span className={getStyleName(styles, 'with-arrow')} />}
          </span>
        )}
        {this.renderPopupMenu()}
      </div>
    );
  }
}
