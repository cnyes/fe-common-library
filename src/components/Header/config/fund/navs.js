/* eslint-disable no-multi-spaces */

const navs = [
  {
    title: '首頁',
    url: 'http://www.cnyes.com/index.htm',
  },
  {
    title: '新聞',
    url: 'http://news.cnyes.com/',
  },
  {
    title: '台股',
    url: 'http://www.cnyes.com/twstock/index.htm',
    catSlug: 'tw_stock',
    leftList: [
      { title: '台指期', url: 'http://www.cnyes.com/twstock/index.htm' },
      { title: '選擇權', url: 'http://www.cnyes.com/twstock/index.htm' },
      { title: '權證', url: 'http://www.cnyes.com/warrant/tw/index.htm' },
      { title: '興櫃', url: 'http://www.cnyes.com/presh/index.htm' },
      { title: '未上市', url: 'http://www.cnyes.com/pre/index.htm' },
    ],
  },
  {
    title: '股市Talk',
    url: 'https://stock.cnyes.com/',
  },
  {
    title: '外匯',
    url: 'http://www.cnyes.com/forex/index.htm',
    catSlug: 'forex',
    leftList: [
      { title: '新聞', url: '/news/cat/forex' },
      { title: '即時報價', url: 'http://traderoom.cnyes.com/global/forex.aspx' },
      { title: '看盤室', url: 'http://www.cnyes.com/forex/forex_index.aspx' },
      { title: '主要匯價', url: 'http://www.cnyes.com/forex/major.aspx' },
      { title: '交叉匯率', url: 'http://www.cnyes.com/forex/forex_list.aspx' },
      { title: '新台幣', url: 'http://www.cnyes.com/forex/twd.aspx' },
      { title: '人民幣', url: 'http://www.cnyes.com/forex/rmb.htm' },
      { title: '日元', url: 'http://www.cnyes.com/forex/jpy.htm' },
      { title: '歐元', url: 'http://www.cnyes.com/forex/eur.htm' },
    ],
  },
  {
    title: '全球市場',
    url: 'http://www.cnyes.com/global/IndexImmediateQuotedPrice/',
    catSlug: 'wd_stock',
    leftList: [
      { title: '國際股', url: 'http://www.cnyes.com/global/IndexImmediateQuotedPrice/' },
      { title: '美股', url: 'http://www.cnyes.com/USASTOCK/index.htm' },
      { title: '港股', url: 'http://www.cnyes.com/hkstock/index.htm' },
      { title: '滬深股', url: 'http://www.cnyes.com/shstock/index.htm' },
      { title: '日股', url: 'http://www.cnyes.com/JP/index.htm' },
      { title: '國際期貨', url: 'http://www.cnyes.com/futures/index.htm' },
      { title: '債券', url: 'http://www.cnyes.com/bond/index.htm' },
      { title: '黃金', url: 'http://www.cnyes.com/gold/index.htm' },
      { title: '全球央行', url: 'http://www.cnyes.com/CentralBank/index.htm' },
      { title: '經濟指標', url: 'http://www.cnyes.com/economy/indicator/Page/schedule.aspx' },
      { title: 'StockQ', url: 'http://money.cnyes.com/StockQ.aspx' },
    ],
  },
  {
    title: '基金',
    url: 'http://fund.cnyes.com/index.htm',
    catSlug: 'fund',
    leftList: [
      { title: '我的基金', url: 'http://fund.cnyes.com/MyFunds.aspx' },
      { title: '基金搜尋', url: 'http://fund.cnyes.com/Search.aspx' },
      { title: '基金排行', url: 'http://fund.cnyes.com/ranking/index.htm' },
      { title: '債券專區', url: 'http://fund.cnyes.com/Fixedincome/' },
      { title: '配息專區', url: 'http://fund.cnyes.com/dividend/index.htm' },
      { title: '研究報告', url: 'http://fund.cnyes.com/report/report/index.htm' },
      { title: '基金交易平台', url: 'https://www.fundsyes.com/?utm_source=cnyes&utm_medium=index_menu' },
    ],
  },
  {
    title: '交易',
    url: '',
    leftList: [
      { title: '基金交易', url: 'https://www.fundsyes.com/Index.aspx?utm_source=cnyes&utm_medium=channel_mainpage' },
    ],
    rightListTitle: '鉅亨基金交易平台',
    rightList: [
      { title: '交易登入', url: 'https://www.fundsyes.com/Login.aspx?utm_source=cnyes&utm_medium=channel_login' },
      {
        title: '免費開戶',
        url: 'https://www.fundsyes.com/launch/fund_index.htm?utm_source=cnyes&utm_medium=channel_newuser',
      },
      {
        title: '最新優惠',
        url: 'https://www.fundsyes.com/Guide/Expenses.aspx?utm_source=cnyes&utm_medium=channel_preferential',
      },
      {
        title: '投資雷達',
        url: 'https://www.fundsyes.com/Tool/SuggestionList.aspx?utm_source=cnyes&utm_medium=channel_radar',
      },
    ],
  },
  {
    title: '理財',
    url: 'http://www.cnyes.com/money/index.htm',
    catSlug: 'tw_money',
    leftList: [
      { title: '銀行服務', url: 'http://www.cnyes.com/money/BankService.aspx' },
      { title: '房貸', url: 'http://cnyes.com/mortgage' },
      { title: '投資分析', url: 'http://www.cnyes.com/money/fundptcalcu.aspx' },
      { title: '民生物價', url: 'http://www.cnyes.com/money/livelihoodPrice.htm' },
      { title: '試算工具', url: 'http://www.cnyes.com/money/BankCalculation.aspx' },
      { title: '理財新聞', url: '/news/cat/tw_money' },
      { title: '理財雜誌', url: 'http://mag.cnyes.com/Column/financial/index.shtml' },
      { title: '固定收益', url: 'http://www.cnyes.com/fixedincome/index.htm' },
      { title: '保險', url: 'http://www.cnyes.com/insurance/index.htm' },
      { title: '租賃', url: 'http://www.cnyes.com/leases/index.htm' },
      { title: '股市地圖', url: 'http://www.cnyes.com/globalinvest/GlobalStock.htm' },
    ],
  },
  {
    title: '海外房產',
    url: 'http://house.cnyes.com/global/',
    catSlug: 'cnyeshouse',
    leftList: [
      { title: '馬來西亞', url: 'http://house.cnyes.com/global/malaysia/default.aspx' },
      { title: '日本', url: 'http://house.cnyes.com/global/japan/' },
      { title: '泰國', url: 'http://house.cnyes.com/global/thailand/' },
      { title: '柬埔寨', url: 'http://house.cnyes.com/global/Cambodia/' },
      { title: '澳洲', url: 'http://house.cnyes.com/global/Australia/' },
      { title: '美國', url: 'http://house.cnyes.com/global/American/' },
      { title: '台灣', url: 'http://house.cnyes.com/' },
      { title: '房地產新聞', url: 'http://house.cnyes.com/News/' },
    ],
  },
  {
    title: '部落格',
    url: 'http://blog.cnyes.com/',
    leftList: [
      { title: '名家專區', url: 'http://blog.cnyes.com/Excerpt.aspx' },
      { title: '人氣排行榜', url: 'http://blog.cnyes.com/SearchBlog.aspx?Purpose=Hit&ga=nav' },
      { title: '最新文章', url: 'http://blog.cnyes.com/WorldFresh.aspx' },
    ],
  },
  {
    title: '主題投資',
    url: 'http://theme.cnyes.com/fund/index.html',
  },
];

export default navs;
