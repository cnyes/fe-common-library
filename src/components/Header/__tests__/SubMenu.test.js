/* eslint-disable import/first,func-names,global-require,prefer-arrow-callback */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import SubMenu from '../SubMenu';
import navs from '../config/fund/navs';
import mockSubMenu from './mocks/mockSubMenu.json';

describe('<SubMenu/>', function() {
  let makeSubject;
  let Link;

  beforeEach(() => {
    jest.resetModules();

    global.__SERVER__ = false;
    global.__CLIENT__ = true;

    Link = jest.fn(() => <div className="mock-link" />);

    makeSubject = params => {
      const props = params || navs[2];

      return mount(<SubMenu {...props} />);
    };
  });

  it('should make a snapshot', () => {
    const subject = makeSubject({
      ...navs[2],
      request: jest.fn(),
    });

    subject.setState({
      list: mockSubMenu.items.data,
    });

    expect(toJson(subject)).toMatchSnapshot();
  });

  describe('render submenu-title', () => {
    it('should be nice if url.length', () => {
      const subject = makeSubject();

      expect(subject.props().url.length).not.toEqual(0);
      expect(subject.find('a.submenu-title').length).toEqual(1);
    });

    it('should be nice if !url.length', () => {
      const subject = makeSubject({
        ...navs[0],
        url: '',
      });

      expect(subject.props().url.length).toEqual(0);
      expect(subject.find('a.submenu-title').length).toEqual(0);
    });
  });

  describe('renderPopupMenu', () => {
    it('should not renderPopupMenu if no left and no right', () => {
      const subject = makeSubject(navs[0]);

      expect(subject.find('.popup').length).toEqual(0);
    });

    it('should be nice if leftList and rightList', () => {
      const subject = makeSubject(navs[7]);

      expect(subject.find('.link-wrapper').length).toEqual(1);
      expect(subject.find('.news-list').length).toEqual(1);

      expect(subject.find('.link-wrapper').html()).toContain('基金交易');
      expect(subject.find('.news-list').html()).toContain('交易登入');
      expect(subject.find('.news-list').html()).toContain('鉅亨基金交易平台');
    });

    it('should be nice if catSlug and requst are both defined', () => {
      const subject = makeSubject({
        ...navs[2],
        request: jest.fn(),
      });

      expect(subject.props().catSlug).toBeDefined();
      expect(subject.props().request).toBeDefined();

      subject.setState({
        list: mockSubMenu.items.data,
      });

      expect(subject.find('.news-list').length).toEqual(1);
      expect(subject.find('.news-list a').length).toEqual(mockSubMenu.items.data.length);
      expect(subject.find('.news-list').html()).toContain('新聞頭條');
    });
  });

  describe('render some link right', () => {
    it('should be nice with props.Link', () => {
      const request = () => {};

      request.get = jest.fn(() => Promise.resolve(mockSubMenu));

      const subject = makeSubject({
        ...navs[4],
        Link,
        request,
      });

      subject.setState({
        list: mockSubMenu.items.data,
      });
      expect(subject.props().Link).toBeDefined();

      expect(subject.find('.popup .link-wrapper .mock-link').length).toEqual(1);
      expect(subject.find('.popup .link-wrapper a').length).toEqual(subject.props().leftList.length - 1);
      expect(subject.find('.popup .news-list .mock-link').length).toEqual(mockSubMenu.items.data.length);
    });

    it('should be nice without props.Link', () => {
      const request = () => {};

      request.get = jest.fn(() => Promise.resolve(mockSubMenu));

      const subject = makeSubject({
        ...navs[4],
        request,
      });

      subject.setState({
        list: mockSubMenu.items.data,
      });
      expect(subject.props().Link).not.toBeDefined();

      expect(subject.find('.popup .link-wrapper .mock-link').length).toEqual(0);
      expect(subject.find('.popup .link-wrapper a').length).toEqual(subject.props().leftList.length);
      expect(subject.find('.popup .news-list .mock-link').length).toEqual(0);
      expect(subject.find('.popup .news-list a').length).toEqual(mockSubMenu.items.data.length);
    });
  });

  describe('load headeline with category', () => {
    it('should load headeline if pros.catSlug is not defined', () => {
      const request = () => {};

      request.get = jest.fn(() => Promise.resolve(mockSubMenu));

      const subject = makeSubject({
        ...navs[1],
        request,
      });

      expect(subject.props().catSlug).not.toBeDefined();
      expect(subject.props().request).toBeDefined();

      subject.simulate('mouseEnter');

      expect(request.get).not.toBeCalled();
    });

    it('should load headeline if pros.catSlug and props.request are both defined', () => {
      const request = () => {};

      request.get = jest.fn(() => Promise.resolve(mockSubMenu));

      const subject = makeSubject({
        ...navs[2],
        request,
      });

      expect(subject.props().catSlug).toBeDefined();
      expect(subject.props().request).toBeDefined();

      subject.simulate('mouseEnter');

      expect(request.get).toBeCalled();
    });
  });
});
