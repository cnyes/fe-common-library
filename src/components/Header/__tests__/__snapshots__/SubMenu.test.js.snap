// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubMenu/> should make a snapshot 1`] = `"<div class=\\"\\"><a href=\\"http://www.cnyes.com/twstock/index.htm\\" class=\\"submenu-title\\"><!-- react-text: 3 -->台股<!-- /react-text --><span class=\\"with-arrow\\"></span></a><div class=\\"popup theme-popup\\"><nav class=\\"link-wrapper theme-link-wrapper\\"><a href=\\"http://www.cnyes.com/twstock/index.htm\\" title=\\"台指期\\">台指期</a><a href=\\"http://www.cnyes.com/twstock/index.htm\\" title=\\"選擇權\\">選擇權</a><a href=\\"http://www.cnyes.com/warrant/tw/index.htm\\" title=\\"權證\\">權證</a><a href=\\"http://www.cnyes.com/presh/index.htm\\" title=\\"興櫃\\">興櫃</a><a href=\\"http://www.cnyes.com/pre/index.htm\\" title=\\"未上市\\">未上市</a></nav><nav class=\\"news-list theme-news-list\\"><h5>新聞頭條</h5><a href=\\"/news/id/3783319\\" title=\\"〈財部拚稅改〉對個人、營利事業影響都很大 須緊盯七大重點\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">〈財部拚稅改〉對個人、營利事業影響都很大 須緊盯七大重點</a><a href=\\"/news/id/3783318\\" title=\\"台股盤後－台積電領跌破季線 多方力抗回測壓力 小跌16點下探9700\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">台股盤後－台積電領跌破季線 多方力抗回測壓力 小跌16點下探9700</a><a href=\\"/news/id/3783271\\" title=\\"高鐵南港轉運站西站明啟用 自強號停靠大增超過百班\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">高鐵南港轉運站西站明啟用 自強號停靠大增超過百班</a><a href=\\"/news/id/3783232\\" title=\\"美系資料中心訂單動能強 信驊Q2營收將再創高\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">美系資料中心訂單動能強 信驊Q2營收將再創高</a><a href=\\"/news/id/3783215\\" title=\\"晶技初估Q1匯損逾8000萬 首季每股稅前賺1.15元\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">晶技初估Q1匯損逾8000萬 首季每股稅前賺1.15元</a><a href=\\"/news/id/3783107\\" title=\\"三大變因 恐成台股上攻絆腳石\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">三大變因 恐成台股上攻絆腳石</a><a href=\\"/news/id/3783183\\" title=\\"雲縣核准六輕許可證   環團靜坐抗議\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">雲縣核准六輕許可證   環團靜坐抗議</a><a href=\\"/news/id/3783136\\" title=\\"夏普2016年財報可達標？戴正吳：有信心\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">夏普2016年財報可達標？戴正吳：有信心</a><a href=\\"/news/id/3783150\\" title=\\"人民幣匯損來襲 弘帆跌出百元俱樂部 股價創一年半新低\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">人民幣匯損來襲 弘帆跌出百元俱樂部 股價創一年半新低</a><a href=\\"/news/id/3783015\\" title=\\"DRAM價格第二季續旺！惟恐壓縮PC、手機、伺服器系統商獲利\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">DRAM價格第二季續旺！惟恐壓縮PC、手機、伺服器系統商獲利</a><a href=\\"/news/id/3783128\\" title=\\"台商前進東協拿什麼打仗？IDC：靠「技術破壞者」優勢\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">台商前進東協拿什麼打仗？IDC：靠「技術破壞者」優勢</a><a href=\\"/news/id/3783130\\" title=\\"鉅橡看旺Q2類載板材料需求 股價逆勢抗震走高\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">鉅橡看旺Q2類載板材料需求 股價逆勢抗震走高</a><a href=\\"/news/id/3783105\\" title=\\"〈大同董監改選〉市場派新大同提抗告明開庭 公司：深表遺憾、委任律師提答辯\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">〈大同董監改選〉市場派新大同提抗告明開庭 公司：深表遺憾、委任律師提答辯</a><a href=\\"/news/id/3783101\\" title=\\"破除資源分散疑慮 發揮鴻海合併綜效 夏普擬重返PC市場\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">破除資源分散疑慮 發揮鴻海合併綜效 夏普擬重返PC市場</a><a href=\\"/news/id/3783102\\" title=\\"台股盤中—新台幣又升+北韓情勢不穩  失守9700點跌破季線\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">台股盤中—新台幣又升+北韓情勢不穩  失守9700點跌破季線</a><a href=\\"/news/id/3783072\\" title=\\"力成買日封測廠 與美光關係更緊密 股價開盤大漲逼近季線\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">力成買日封測廠 與美光關係更緊密 股價開盤大漲逼近季線</a><a href=\\"/news/id/3783063\\" title=\\"國泰金：經濟、就業樂觀指數連3月攀高 惟僅5.2%民眾估今年GDP會破2\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">國泰金：經濟、就業樂觀指數連3月攀高 惟僅5.2%民眾估今年GDP會破2</a><a href=\\"/news/id/3783035\\" title=\\"MLCC報價7年來首調漲 華新科強彈 凱美、智寶續拉第3根漲停\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">MLCC報價7年來首調漲 華新科強彈 凱美、智寶續拉第3根漲停</a><a href=\\"/news/id/3783059\\" title=\\"台股痛失「千金」 精測股價再重挫跌破千元 波段市值蒸發逾百億\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">台股痛失「千金」 精測股價再重挫跌破千元 波段市值蒸發逾百億</a><a href=\\"/news/id/3782849\\" title=\\"新世代台商逐鹿東協 掌握6大投資要素最為關鍵\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">新世代台商逐鹿東協 掌握6大投資要素最為關鍵</a><a href=\\"/news/id/3783012\\" title=\\"晶華首次攜手好市多 開賣冷凍禮盒\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">晶華首次攜手好市多 開賣冷凍禮盒</a><a href=\\"/news/id/3782008\\" title=\\"鉅亨早報─中國第1季GDP、日央行行長談話\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">鉅亨早報─中國第1季GDP、日央行行長談話</a><a href=\\"/news/id/3782983\\" title=\\"台灣主要報紙頭條要聞 2017年4月17日\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">台灣主要報紙頭條要聞 2017年4月17日</a><a href=\\"/news/id/3782987\\" title=\\"台股盤前─報稅季、匯損、北韓局勢多方消息衝擊 大盤盤弱壓力大\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">台股盤前─報稅季、匯損、北韓局勢多方消息衝擊 大盤盤弱壓力大</a><a href=\\"/news/id/3782776\\" title=\\"全球戰情緊張壓抑台股短線觀望適時汰弱留強\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">全球戰情緊張壓抑台股短線觀望適時汰弱留強</a><a href=\\"/news/id/3782750\\" title=\\"〈北韓若生戰事〉對雙D市場影響更甚於台灣921地震  「暫停報價」、「一日多市」恐再現\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">〈北韓若生戰事〉對雙D市場影響更甚於台灣921地震  「暫停報價」、「一日多市」恐再現</a><a href=\\"/news/id/3782714\\" title=\\"〈北韓若生戰事〉供不應求問題擴大 面板價格恐再漲\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">〈北韓若生戰事〉供不應求問題擴大 面板價格恐再漲</a><a href=\\"/news/id/3782649\\" title=\\"〈名師解盤〉台股上有鍋蓋下有鐵板 只怕狂人失控遭池魚之殃\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">〈名師解盤〉台股上有鍋蓋下有鐵板 只怕狂人失控遭池魚之殃</a><a href=\\"/news/id/3782690\\" title=\\"〈北韓若生戰事〉若有進一步挑釁 買方恐生恐慌 再推升DRAM價格漲勢\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">〈北韓若生戰事〉若有進一步挑釁 買方恐生恐慌 再推升DRAM價格漲勢</a><a href=\\"/news/id/3782467\\" title=\\"供不應求！DrameXchange：第二季DRAM報價喊漲12.5%\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">供不應求！DrameXchange：第二季DRAM報價喊漲12.5%</a></nav></div></div>"`;
